lib/appsignal/oban.ex

```diff
@@ -0,0 +1,172 @@
+      [:oban, :engine, :insert_job, :exception] => &__MODULE__.oban_insert_job_exception/4
+    }
+
+    for {event, fun} <- handlers do
+      case :telemetry.attach({__MODULE__, event}, event, fun, :ok) do
+        :ok ->
+          _ = Appsignal.IntegrationLogger.debug("Appsignal.Oban attached to #{inspect(event)}")
+
+          :ok
+
+        {:error, _} = error ->
+          Logger.warn("Appsignal.Oban not attached to #{inspect(event)}: #{inspect(error)}")
+
+          error
+      end
+    end
+  end
+
+  def oban_job_start(
+        _event,
+        _measurements,
+        %{job: job},
+        _config
+      ) do
+    span = @tracer.create_span("oban")
+
+    span
+    |> @span.set_name(to_string(job.worker))
+    |> @span.set_sample_data("params", job.args)
+    |> @span.set_attribute("id", job.id)
+    |> @span.set_attribute("queue", to_string(job.queue))
+    |> @span.set_attribute("attempt", job.attempt)
+    |> @span.set_attribute("priority", job.priority)
+    |> @span.set_attribute("appsignal:category", "job.oban")
+
+    for {key, value} <- job.tags do
+      @span.set_attribute(span, "job_tag_#{key}", value)
+    end
+
+    increment_job_start_counter(job)
+
+    add_job_queue_time_value(job)
+  end
+
+  def oban_job_stop(
+        _event,
+        %{duration: duration},
+        %{state: state, result: result, job: job},
+        _config
+      ) do
+    @tracer.current_span()
+    |> @span.set_attribute("state", to_string(state))
+    |> @span.set_attribute("result", inspect(result))
+    |> @tracer.close_span()
+
+    increment_job_stop_counter(job, state)
+
+    add_job_duration_value(job, duration, state)
+  end
+
+  def oban_job_exception(_event, %{duration: duration}, metadata, _config) do
+    span =
+      @tracer.current_span()
+      |> @span.set_attribute("state", to_string(metadata[:state]))
+      |> @span.set_attribute("worker", to_string(metadata[:job].worker))
+
+    if Map.has_key?(metadata, :kind) and Map.has_key?(metadata, :reason) and
+         Map.has_key?(metadata, :stacktrace) do
+      @span.add_error(span, metadata[:kind], metadata[:reason], metadata[:stacktrace])
+    end
+
+    @tracer.close_span(span)
+
+    increment_job_stop_counter(metadata[:job], metadata[:state])
+
+    add_job_duration_value(metadata[:job], duration, metadata[:state])
+  end
+
+  def oban_insert_job_start(_event, _measurements, metadata, _config) do
+    span = @tracer.create_span("oban", @tracer.current_span)
+
+    @span.set_attribute(span, "appsignal:category", "insert_job.oban")
+
+    case metadata[:changeset] do
+      %{changes: %{worker: worker}} ->
+        @span.set_name(span, "Insert job (#{worker})")
+
+      _ ->
+        @span.set_name(span, "Insert job")
+    end
+  end
+
+  def oban_insert_job_stop(_event, _measurements, metadata, _config) do
+    case metadata[:changeset] do
+      %{changes: %{worker: worker}} ->
+        @appsignal.increment_counter("oban_job_insert", 1, %{worker: to_string(worker)})
```

Comment:
```suggestion
        @appsignal.increment_counter("oban_job_queued", 1, %{worker: to_string(worker)})
```

Queued would be a better term here, right?
I'm also on the fence if it's interesting data to see how many jobs were queued, if you can also see how many jobs were processed.

Reply:
The way I thought about it was that it could be interesting to contrast this with the job start/stop -- if your queues are saturated, then the jobs inserted will be more than the jobs processed. Also as a "sanity check" of sorts: "are my queues not doing anything because no messages are being queued, or because the queues broke?"

I'd consider removing them if you don't see the value. The name change sounds good to me.

Reply:
If I compare it to Sidekiq again, there we collect the total number of queued jobs for all queues, which is a more actionable metric. You don't have to compare it against other metrics to see no host is processing a certain queue or it's handling them too slow.

Because Oban stores it in postgres I'd assume we can also fetch those queue counts that way. We can do that at a later time. We don't have to worry about updating the dashboard config anymore, we have an updater for that.

> Also as a "sanity check" of sorts: "are my queues not doing anything because no messages are being queued

I haven't heard this request before. I guess that's one purpose for it, but my thoughts go to testing to see if the job is queued or not.

Reply:
Sounds good to me. Will remove this metric.